File: leetcode-solutions/Python/Array/14_LongestCommonPrefix.py
#ID: 14
#Problem: Longest Common Prefix
#Difficulty: Easy
#Link: https://leetcode.com/problems/longest-common-prefix/

from typing import List

def longestCommonPrefix(strs: List[str]) -> str:
    #Storing first word
    prefix = strs[0]

    # Compare prefix against each remaining word
    for word in strs[1:]:
        # Shorten prefix from the end until current word starts with it
        while not word.startswith(prefix):
            #slicing [start:end] (-1 start from end of word)
            prefix = prefix[:-1] #removed last char
        
            #Return empty string for no common prefix
            if prefix == "":
                return ""      
    return prefix


# Example test cases
if __name__ == "__main__":
    print(longestCommonPrefix(["flower","flow", "flight"])) #Output: "fl"
    print(longestCommonPrefix(["dog","cat","mouse"]))       #Output: ""

File: leetcode-solutions/C++/Array/14_LongestCommonPrefix.cpp
// ID: 14
// Problem: Longest Common Prefix
// Difficulty: Easy
// Link: https://leetcode.com/problems/longest-common-prefix/
// Approach: Iteration with for loop over words + while loop to shrink 

# include <iostream>
# include <vector>

using namespace std;

 string longestCommonPrefix(vector<string>& strs) {
        // Empty string returns nothing
        if (strs.empty()) {
            return "";
        }

        // Taking  first words as prefix in list
        string prefix = strs[0];

        for(int i = 1; i < strs.size(); ++i){
            // Remove last letter if current word does not match prefix
            while (strs[i].find(prefix) != 0){
                prefix = prefix.substr(0, prefix.size() - 1 );
                if (prefix.empty()){
                    return "";
                }
            }
        }
        return prefix;
    }


//Test Case 
int main(){

    // Case 1 
    vector<string> words = {"flower", "flow", "flight"};
    cout << "Output: " << longestCommonPrefix(words) << endl;

    // Case 2
    vector<string> words2 = {"dog", "log", "car", "racecar"};
    cout << "Output: " << longestCommonPrefix(words2) << endl;

    return 0;
}


